src/main/scala/com/databricks/spark/xml/parsers/StaxXmlParser.scala

```diff
@@ -211,13 +217,17 @@ private[xml] object StaxXmlParser {
         valuesMap + (options.valueTag -> v)
       case _ => Map.empty
     }
-    // The fields are sorted so `TreeMap` is used.
-    val convertedValuesMap = convertValues(valuesMap, schema)
-    val row = TreeMap((fields ++ convertedValuesMap).toSeq : _*).values.toSeq
+
+    // Here we merge both to a row.
+    val valuesMap = fieldsMap ++ attributesMap
+    valuesMap.foreach { case (f, v) =>
+      val nameToIndex = schema.map(_.name).zipWithIndex.toMap
+      nameToIndex.get(f).foreach(row.update(_, v))
```

Comment:
I think with an assignment to the row it'd be easier to spot what's happening here...